File: Semester 1/Class14_Computer_Contest_Lv_2_Notes_and_Homework_Dec_11/lvl2_hw_sol/lvl2_queens_cant_attack_me.py
import sys
input = sys.stdin.readline
n, m = map(int, input().split())
row, col, left, right = [True]*n, [True]*n, [True]*(2*n-1), [True]*(2*n-1)
for _ in range(m):
    r, c = map(int, input().split())
    r, c = r-1, c-1
    row[r] = col[c] = left[r+c] = right[r+n-1-c] = False
ans = 0
for r in range(n):
    for c in range(n):
        if row[r] and col[c] and left[r+c] and right[r+n-1-c]:
            ans+=1
print(ans)

File: Semester 1/Class14_Computer_Contest_Lv_2_Notes_and_Homework_Dec_11/lvl2_hw_sol/lvl2_queens_cant_attack_me.java
import java.util.*;
import java.io.*;
public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException{	
		int n = readInt(), m = readInt();
		boolean [] row = new boolean[n], col = new boolean[n], left = new boolean[2*n-1], right = new boolean[2*n-1];
		for(int i=1; i<=m; i++) {
			int r = readInt() - 1, c = readInt() - 1;
			row[r] = col[c] = left[r+c] = right[r - c + n - 1] = true;
		}
		int ans=0;
		for(int r=0; r<n; r++) {
			for(int c=0; c<n; c++) {
				if(!row[r] && !col[c] && !left[r+c] && !right[r-c+n-1]) ans++;
			}
		}
		System.out.println(ans);
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}